tests/raftstore/test_snap.rs

```diff
@@ -344,3 +346,89 @@ fn test_node_stale_snap() {
     cluster.clear_send_filters();
     must_get_equal(&cluster.get_engine(3), b"k3", b"v3");
 }
+
+/// Pause Snap and wait till first append message arrives.
+pub struct SnapshotApendFilter {
+    stale: AtomicBool,
+    pending_msg: Mutex<Vec<Msg>>,
+    notifier: Mutex<Sender<()>>,
+}
+
+impl SnapshotApendFilter {
+    pub fn new(notifier: Sender<()>) -> SnapshotApendFilter {
+        SnapshotApendFilter {
+            stale: AtomicBool::new(false),
+            pending_msg: Mutex::new(vec![]),
+            notifier: Mutex::new(notifier),
+        }
+    }
+}
+
+impl Filter<Msg> for SnapshotApendFilter {
+    fn before(&self, msgs: &mut Vec<Msg>) -> Result<()> {
+        if self.stale.load(Ordering::Relaxed) {
+            return Ok(());
+        }
+        let mut to_send = vec![];
+        let mut pending_msg = self.pending_msg.lock().unwrap();
+        let mut stale = false;
+        for m in msgs.drain(..) {
+            let mut should_collect = false;
+            if let Msg::RaftMessage(ref msg) = m {
+                should_collect = !stale &&
+                                 msg.get_message().get_msg_type() == MessageType::MsgSnapshot;
+                stale = !pending_msg.is_empty() &&
+                        msg.get_message().get_msg_type() == MessageType::MsgAppend;
+            }
+            if should_collect {
+                pending_msg.push(m);
+                self.notifier.lock().unwrap().send(()).unwrap();
+            } else {
+                if stale {
+                    to_send.extend(pending_msg.drain(..));
+                }
+                to_send.push(m);
+            }
+        }
+        self.stale.store(stale, Ordering::SeqCst);
+        msgs.extend(to_send);
+        Ok(())
+    }
+}
+
+fn test_snapshot_with_append<T: Simulator>(cluster: &mut Cluster<T>) {
+    // truncate the log quickly so that we can force sending snapshot.
+    cluster.cfg.raft_store.raft_log_gc_tick_interval = 20;
+    cluster.cfg.raft_store.raft_log_gc_limit = 2;
+    cluster.cfg.raft_store.snap_mgr_gc_tick_interval = 50;
+    cluster.cfg.raft_store.snap_gc_timeout = 2;
+
+    let pd_client = cluster.pd_client.clone();
+    // Disable default max peer count check.
+    pd_client.disable_default_rule();
+    let r1 = cluster.run_conf_change();
+    pd_client.must_add_peer(r1, new_peer(2, 2));
+    pd_client.must_add_peer(r1, new_peer(3, 3));
+
+    let (tx, rx) = mpsc::channel();
+    cluster.sim.wl().add_recv_filter(4, box SnapshotApendFilter::new(tx));
+    pd_client.add_peer(r1, new_peer(4, 4));
+    rx.recv_timeout(Duration::from_secs(3)).unwrap();
+    cluster.must_put(b"k1", b"v1");
```

Comment:
Only meta data is included in the snapshot. Append entries are (k1, v1) and (k2, v2).